src/main/java/io/quarkus/bot/ApproveWorkflow.java

```diff
@@ -0,0 +1,214 @@
+                        approval.shouldNotApprove = true;
+                    }
+                }
+            }
+        }
+    }
+
+    public static boolean matchRuleFromChangedFiles(GHPullRequest pullRequest,
+            QuarkusGitHubBotConfigFile.WorkflowApprovalCondition rule) {
+        // for now, we only use the files but we could also use the other rules at some point
+        if (rule == null) {
+            return false;
+        }
+
+        boolean matches = false;
+
+        PullRequestFilesMatcher prMatcher = new PullRequestFilesMatcher(pullRequest);
+        if (matchDirectories(prMatcher, rule)) {
+            matches = true;
+        } else if (matchFiles(prMatcher, rule)) {
+            matches = true;
+        }
+
+        return matches;
+    }
+
+    private static boolean matchDirectories(PullRequestFilesMatcher prMatcher,
+            QuarkusGitHubBotConfigFile.WorkflowApprovalCondition rule) {
+        if (rule.directories == null || rule.directories.isEmpty()) {
+            return false;
+        }
+        if (prMatcher.changedFilesMatchDirectory(rule.directories)) {
+            return true;
+        }
+        return false;
+    }
+
+    private static boolean matchFiles(PullRequestFilesMatcher prMatcher,
+            QuarkusGitHubBotConfigFile.WorkflowApprovalCondition rule) {
+        if (rule.files == null || rule.files.isEmpty()) {
+            return false;
+        }
+        if (prMatcher.changedFilesMatchFile(rule.files)) {
+            return true;
+        }
+        return false;
+    }
+
+    private boolean matchRuleForUser(GHRepositoryStatistics.ContributorStats stats,
+            QuarkusGitHubBotConfigFile.WorkflowApprovalCondition rule) {
+        if (rule == null || stats == null) {
+            return false;
+        }
+
+        if (rule.users == null) {
+            return false;
+        }
+
+        if (rule.users.minContributions != null && stats.getTotal() >= rule.users.minContributions) {
+            return true;
+        }
+
+        // We can add more rules here, for example how long the user has been contributing
+
+        return false;
+    }
+
+    private GHRepositoryStatistics.ContributorStats getStatsForUser(GHEventPayload.WorkflowRun workflowPayload) {
+        if (workflowPayload.getSender().getLogin() != null) {
+            try {
+                GHRepositoryStatistics statistics = workflowPayload.getRepository().getStatistics();
+                if (statistics != null) {
+                    PagedIterable<GHRepositoryStatistics.ContributorStats> contributors = statistics.getContributorStats();
+                    for (GHRepositoryStatistics.ContributorStats contributor : contributors) {
+                        if (workflowPayload.getSender().getLogin().equals(contributor.getAuthor().getLogin())) {
+                            return contributor;
+                        }
+                    }
+                }
```

Comment:
I'm worried about this. With a repository the size of Quarkus, it's going to trigger quite a lot of queries if you're not lucky. Also this triggers a statistics computation on the GH side if not already cached and it takes quite a lot of time (and I have no idea of the refresh rate).

I think we need a cache for this with a list of the already validated user logins.

Reply:
Good point. I'll update with some caching. How long-lived is the app process? Will a static map do, or will that be reloaded too often?

Reply:
For now, I think a static map would do. My guess is that at some point we could have this stored into a repository file and have commands to actually add someone to this file. But that's something for another day I would say.

Reply:
Oh just to be clear: you need to add to the Map only the ones who qualifies for it to work if you're going fully static. Because once they qualify, we can be sure they always qualify.

Also, I'm wondering if we should use https://docs.github.com/en/rest/search#search-issues-and-pull-requests instead. With this API, we can search by author IIRC (it's the API used in the pull requests page so `author:someone` should work). And you could count the pull requests merged instead of relying on the statistics.

Reply:
Something like `is:pr is:merged author:gsmet ` and then you would define the pageSize to the limit, call `nextPage()` on the `PagedIterator` and check if you have at least limit elements.

I think that's probably the most efficient way to check the number of contributions. Also better to base this thing on the number of PR merged rather than the number of commits so it's all win.

Reply:
I've implemented caching based on the stats, so we can see what it looks like. 

I think no matter which way we go, we'd need to cache a count of contributions for each user, not just a yes/no per user, because the rules could change. For example, if we say a user needs 5 contributions, and then we realise that bar is too low and all sorts of dubious stuff is getting through so we change the bar to 20, we don't want all the pre-approved people to stay pre-approved. The whole reason we changed the rule was to de-approve them. :) It would be similar in the opposite direction, if we realised we had the bar too high and wanted to lower it to reduce admin load, we'd want that change to apply with immediate effect to people who previously hadn't qualified. And it's so easy to cache the count I don't think there's any particular downside.

Whether we do repository stats or query for merged PRs is a harder call, I think. There’s some advantages to keeping the stats:

- It’s already written, I am lazy, and mocking against the github api is tedious :) 
- It allows us to extend to add new rules like “how long has this person been contributing", which is information that's already in the stats
- In the stats, ‘total’ is an instance /top-level json variable so there’s no further API requests to get it
- It’s exactly the information we need, so it’s possible we get the benefit of github's server-side caching for it in a way we wouldn’t if we tried to assemble the information ourself (https://api.github.com/repos/quarkusio/quarkus/contributors appears to be near-instantaneous). Asking for the PRs for a user would mean a call per user, whereas with the repository stats a smaller number of calls that would allow us to cache the stats for every user.

On the other hand, I think the point about merged PRs being more relevant than contributions is a good one. So I don't know. 



By definition, most users will be ones we haven’t seen before, or they’d already be committers. 

One complication is that cache misses are somewhat likely (any first-time contributor), and we will crawl the contributor table for any first time contributor. The silver lining is that we will refresh the cache while doing this, and the number of first-time contributors should be low. 

// In an ideal world, we would use the paged iterable, and cache individual contributors as we traverse it
// We can't do that until we have https://github.com/quarkusio/quarkus/pull/27488/ ...

Reply:
One other argument in favour of doing the request for each user is that what I had to implement was kind of ugly, with two levels of caching. In an ideal world, we would fetch the paged iterable of stats, and cache individual contributors as we traverse it. That way we might entirely warm the cache after we retrieve just one or two users. However, we can't do that until we have https://github.com/quarkusio/quarkus/pull/27488/ in our Quarkus.

Without the ability to manually populate the cache, I ended up doing one cache for individual users, and a second cache for the results of fetching the repository stats. The first cache is perhaps redundant, but I left it in since it's semantically closer to what we *actually* want to cache.